x/evm/keeper/keeper.go

```diff
@@ -339,3 +341,62 @@ func (k Keeper) DeleteHeaderHash(ctx sdk.Context, height uint64) {
 	store := ctx.KVStore(k.storeKey)
 	store.Delete(types.GetHeaderHashKey(height))
 }
+
+func (k *Keeper) AddPreinstalls(ctx sdk.Context, preinstalls []types.Preinstall) error {
+	for _, preinstall := range preinstalls {
+		address := common.HexToAddress(preinstall.Address)
+		accAddress := sdk.AccAddress(address.Bytes())
+
+		if len(preinstall.Code) == 0 {
+			return errorsmod.Wrapf(types.ErrInvalidPreinstall, "preinstall %s has no code", preinstall.Address)
+		}
+
+		// check that the address does not conflict with the precompiles
+		cfg, err := k.EVMBlockConfig(ctx, k.ChainID())
+		if err != nil {
+			return err
+		}
+		for _, fn := range k.customContractFns {
+			c := fn(ctx, cfg.Rules)
+			if address == c.Address() {
+				return errorsmod.Wrapf(types.ErrInvalidPreinstall, "preinstall %s already exists as a precompile", preinstall.Address)
+			}
+		}
+
+		codeHash := crypto.Keccak256Hash(common.FromHex(preinstall.Code))
+		codeHashBytes := codeHash.Bytes()
+		if types.IsEmptyCodeHash(codeHashBytes) {
+			return errorsmod.Wrapf(types.ErrInvalidPreinstall, "preinstall %s has empty code hash", preinstall.Address)
```

Comment:
Should we also print the preinstall name to make it more readable?

Reply:
sure